### ISSUE TITLE
Test experiments should have limited runtime

### ISSUE THREAD

**Issue by zesch** [dkpro/dkpro-tc#257]
I have been running a lot of tests lately and the test experiments run quite a while.
We should limit the run-time for each test, e.g. by removing some of the test data.

**Comment by Horsmann** [comment]
Test of the example-project run now a bit less than 10 minutes on my machine.

I deleted quite a lot training data. Many tests had for no clear reason SVM and NaiveBayes as ML classifier. I assumed these were copy/paste artifacts. Same for different feature configurations. I furthermore removed the N for the Ngram features down from 1000 to 100 etc.

**Comment by Horsmann** [comment]
A further option for speeding things even more up would be to either CV or either TrainTest. Atm this is seems to arbitrary. Some test both, some only CV and others only TrainTest.

Any thoughts on this? I would tend to go on TrainTest and only in one case on CV to have it covered.

**Comment by daxenberger** [comment]
I think we should have one TrainTest and one CV example and test case for each feature and each learning mode. So far, this has not been implemented consequently, I guess.

**Comment by Horsmann** [comment]
The most tests seemed to have a lot copy / paste artifacts such as feature ablation tests, several classifier (most the time only SVM and NavBay). This slowed things down additionally especially in the CV mode.

So you want one huge test with all TC features that iterates all ML classifier? While the actual "api examples" run than only in TrainTest?

**Comment by daxenberger** [comment]
> So you want one huge test with all TC features that iterates all ML classifier? 
> While the actual "api  examples" run than only in TrainTest?

Nope. I suggest to have for each feature mode (document, unit, sequence, pair) and for each learning mode (single-label, multi-label, regression) one demo experiment and consequently one test case, which ensures that this configuration can also be executed (we introduced such tests to make sure we note when a modification breaks any of the several experiment setups as we do not have low-level unit tests for everything).

**Comment by Horsmann** [comment]
In total my changes saved 25 minutes Jenkins build-time. It still takes 30 minutes for the Java experiments and 25 for the Groovy experiments (untouched).

**Comment by zesch** [comment]
That is already quite good.
For further savings we might want to ignore some tests that are not testing any new combination of learning/mode feature mode (if we still have such duplication in the tests).

Another optimization would be to use NaiveBayes as the default not SMO, which should be much faster.

**Comment by Horsmann** [comment]
I think the tests do not really have to use all those reports? They also take some time?
We have two goals here which is probably another issue 
1) documentation / how-to-use X
2) integration tests

The former probably should have the reports the latter should not - at least not in every single test.

**Comment by daxenberger** [comment]
> I think the tests do not really have to use all those reports? They also take some time?
> We have two goals here which is probably another issue 
> 1) documentation / how-to-use X
> 2) integration tests
> The former probably should have the reports the latter should not - at least not in every single test.

Reports should be tested via integration as well, but only once, I agree.

**Comment by Horsmann** [comment]
Ok
- I set Naive Bayes as classifier in case the test is not about testing specific ML adapters 
- Added a package with two new tests dedicated to reports
- Outcommented all reports unless the demo used some specific tests (e.g. liblinear)

This should have made things even faster. Lets see what Jenkins says.

**Comment by Horsmann** [comment]
The change was not very effective, the example module still needs 30 mins. 

The next chance for improvements would be more consolidated tests. I assume loading the entire CV and TT Lab-Stuff anew in the background drains also some CPU time. It would probably better to test more things in a single test - for the Jenkins runtime at least. Readability might suffer and it wouldn't be clear anymore what a test is testing.

What you could consolidate are probably the ML classifier variation. e.g. one test case that tests the most "simple" classifier with their default configuration. I think there are still some redundant tests left that differ only in the data that is used, but don't test different TC components.

**Comment by zesch** [comment]
Do we know which task now eats up the most time?
Maybe it is now preprocessing or feature extraction and not classification.

2015-09-24 16:38 GMT+02:00 Tobias Horsmann notifications@github.com:

**Comment by Horsmann** [comment]
The following 3 demos seems obsolete 
de.tudarmstadt.ukp.dkpro.tc.examples.single.document
- SimpleDkproTCReaderDemo
- TwentyNewsgroupsStandardReader
- TwitterSentimentDemo

Would someone miss them?

Attached the runtimes on my local machine for each test - the AblationTest is stuck at snail-speed
![runtime](https://cloud.githubusercontent.com/assets/8234699/10076845/d7bebf50-62dd-11e5-80a5-06880f6e65bf.png)

**Comment by daxenberger** [comment]
> The following 3 demos seems obsolete (...)
> SimpleDkproTCReaderDemo
> TwentyNewsgroupsStandardReader
> TwitterSentimentDemo
> 
> Would someone miss them?

Yes, at least TwitterSentimentDemo is the reference demo we used in the DKPro TC ACL demo. No need for a test, though.
Not sure about the other ones.

**Comment by zesch** [comment]
I don't really know what was the purpose of the SimpleDkproTCReaderDemo.
Probably it was introduced when we added the (now somewhat) obsolete
TCReader interface and base classes.

TwentyNewsgroupsStandardReader is definitely needed.
Maybe the name is not optimal, but this is the only demo we have which uses
a standard Core reader and a specialized AE to set the TCO annotation.
This should also keep its test (if it has one :)

-Torsten

2015-09-24 17:23 GMT+02:00 Johannes Daxenberger notifications@github.com:

**Comment by Horsmann** [comment]
@daxenberger this implies the Twitter training data should stay complete, too? I actually deleted some of it. I can restore it and simply remove the Demo from the TestCases.

**Comment by zesch** [comment]
The amount of data for the Twitter demo should be enough to showcase what
is going on.
The current amount might not be enough for that.
So, yes. Please restore the data and remove the demo from test cases.

2015-09-24 19:43 GMT+02:00 Tobias Horsmann notifications@github.com:

**Comment by daxenberger** [comment]
> @daxenberger this implies the Twitter training data should stay complete, too?

Yes, thanks.

**Comment by zesch** [comment]
We have reduced run-time quite a bit.
Please re-open if you feel that this hasn't been properly resolved yet.
